main/lsp/requests/rename.cc

```diff
@@ -236,32 +285,17 @@ void RenameTask::getRenameEdits(LSPTypecheckerDelegate &typechecker, core::Symbo
     vector<core::SymbolRef> symbolsToRename;
     if (symbolData->isMethod()) {
         renamer = make_unique<MethodRenamer>(gs, config, originalName, newName);
-        // We have to check for methods as part of a class hierarchy: Follow superClass() links till we find the root;
-        // then find the full tree; then look for methods with the same name as ours; then find all references to all
-        // those methods and rename them.
-        auto symbolClass = symbolData->enclosingClass(gs);
-
-        // We have to be careful to follow superclass links only as long as we find a method that `symbol` overrides.
-        // Otherwise we will find unrelated methods and rename them even though they don't need to be (see the
-        // method_class_hierarchy test case for an example).
-        auto root = findRootClassWithMethod(gs, symbolClass, symbolData->name);
-
-        auto subclasses = getSubclasses(typechecker, root);
-
-        // find the target method definition in each subclass
-        for (auto c : subclasses) {
-            auto classSymbol = c.data(gs);
-            auto member = classSymbol->findMember(gs, symbolData->name);
-            if (member.exists()) {
-                symbolsToRename.push_back(member);
-            }
-        }
+        addSubclassRelatedMethods(typechecker, symbol, symbolsToRename);
     } else {
         renamer = make_unique<ConstRenamer>(gs, config, originalName, newName);
         symbolsToRename.push_back(symbol);
     }
 
-    for (auto sym : symbolsToRename) {
+    // We use symbolsToRename as a queue -- the loop body may add more symbols that need to be renamed, which is why we
+    // don't use an iterator here.
+    for (int i = 0; i < symbolsToRename.size(); i++) {
```

Comment:
Ideally, this would be a `std::deque` and we loop `while (!deque.empty())`. But we also need to make sure we don't process duplicate symbols.

Should we build a simple class wrapper that provides this functionality (e.g., with a `tryEnqueue` that returns a `bool` indicating if the thing was enqueued)? It could contain a simple `UnorderedSet<SymbolRef>` to track symbols that have already been enqueued.

Reply:
Yes this is a good idea, will do.